cdi/src/main/java/org/hibernate/validator/cdi/internal/ValidateableBeanFilter.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.cdi.internal;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Executable;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import javax.validation.Constraint;
+import javax.validation.Valid;
+import javax.validation.executable.ValidateOnExecution;
+
+import org.hibernate.validator.internal.util.classhierarchy.ClassHierarchyHelper;
+
+/**
+ * A filter that checks if the passed class has any of annotaitons:
+ * <ul>
+ * <li>{@link Valid}</li>
+ * <li>{@link Constraint}</li>
+ * <li>{@link ValidateOnExecution}</li>
+ * </ul>
+ * <p>
+ * on a type level, or on any member, or any parameter of any member.
+ * The annotation may be applied as a meta-annotation on any annotation considered
+ *
+ * @author Marko Bekhta
+ */
+public class ValidateableBeanFilter implements Predicate<Class<?>> {
+
+	private static final Set<Class<? extends Annotation>> ALLOWED_ANNOTATIONS = Collections.unmodifiableSet(
+			new HashSet<>( Arrays.asList( Valid.class, Constraint.class, ValidateOnExecution.class ) )
+	);
+
+	@Override
+	public boolean test(Class<?> type) {
+		return ClassHierarchyHelper.getHierarchy( type ).parallelStream()
+				.anyMatch( this::isAnnotationPresentAnywhere );
+	}
+
+	private boolean isAnnotationPresentAnywhere(Class<?> clazz) {
+		Set<Annotation> processedAnnotations = new HashSet<>();
+		// 1. is present on a type level:
+		return Arrays.stream( clazz.getDeclaredAnnotations() )
+				.anyMatch( annotation -> isAnnotationAllowed( annotation, processedAnnotations ) )
+				// 2. or on a field level
+				|| Arrays.stream( clazz.getDeclaredFields() )
+				.flatMap( field -> Arrays.stream( field.getDeclaredAnnotations() ) )
+				.anyMatch( annotation -> isAnnotationAllowed( annotation, processedAnnotations ) )
+				// 3. or on any executable
+				// 3.1 constructors:
+				|| Arrays.stream( clazz.getDeclaredConstructors() )
+				.flatMap( this::fromExecutable )
+				.anyMatch( annotation -> isAnnotationAllowed( annotation, processedAnnotations ) )
+				// 3.2 methods:
+				|| Arrays.stream( clazz.getDeclaredMethods() )
+				.flatMap( this::fromExecutable )
+				.anyMatch( annotation -> isAnnotationAllowed( annotation, processedAnnotations ) );
```

Comment:
I know you love streams and lambdas very much :) but... we are trying to avoid them when there is no clear benefit as there is a cost to it.

In this case, I think a bunch of for loops would be more readable, especially since the `return` will serve as a `break`.